### PR TITLE
Implement Mutation- and BorrowOfLayoutConstrainedField in thir-unsafeck

### PR DESCRIPTION
**Pull request by FabianWolff**
Since nobody has so far claimed Mutation- and BorrowOfLayoutConstrainedField in rust-lang/project-thir-unsafeck#7, I have taken the liberty of implementing them in thir-unsafeck.

r? @LeSeulArtichaut

### CODE REVIEW

**Comment by LeSeulArtichaut** [comment]
Thanks for your help on this! I'll assign this to you in the tracking issue, if you want to implement something else in the future feel free to just drop a comment on the tracking issue, just to make sure that there aren't two people working on the same thing. With that said, the last few unsafe ops to implement are the trickiest, so having multiple eyes on them is a good thing.

I don't have time to do a proper review today, but I think I can already tell that this PR doesn't cover references that arise from patterns (e.g. `ref mut` patterns)? *It doesn't need to be implemented here*, but that's something you can implement if you want additional work.

Also cc @nikomatsakis, if you want to take a look at this as well

**Comment by FabianWolff** [comment]
Thanks for your quick response! I have added some code (and a test case) now that deals with by-reference bindings in `match` arms.

**Comment by goffrie** [comment]
The title of this PR confused me, but you probably didn't mean thir-borrowck :p

**Comment by FabianWolff** [comment]
> The title of this PR confused me, but you probably didn't mean thir-borrowck :p

Fixed, thanks for catching this! Thinking about all the different borrow kinds etc. apparently got me confused...

File: compiler/rustc_mir_build/src/check_unsafety.rs [hunk 1]
```diff
@@ -246,11 +248,135 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
...
+    fn visit_pat(&mut self, pat: &Pat<'tcx>) {
+        match &*pat.kind {
+            PatKind::Variant { .. } | PatKind::Leaf { .. } => {
+                if let ty::Adt(adt_def, ..) = pat.ty.kind() {
+                    if (Bound::Unbounded, Bound::Unbounded)
+                        != self.tcx.layout_scalar_valid_range(adt_def.did)
+                    {
+                        check_for_ref_bindings(self, pat);
+
+                        fn check_for_ref_bindings<'a, 'tcx>(
```
Comment (LeSeulArtichaut) [inline compiler/rustc_mir_build/src/check_unsafety.rs:312]: I don't see why does this only recurses on certain kinds of patterns, what am I missing? Also, it's a bit of a shame to re-implement visiting, and we visit some patterns twice. Can we implement this by having a field in `UnsafetyVisitor` track when we are in a layout-restricted-type pattern, and let `walk_pat` do its thing?

File: compiler/rustc_mir_build/src/check_unsafety.rs [hunk 2]
```diff
@@ -246,11 +248,135 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
                 // Unsafe blocks can be used in closures, make sure to take it into account
                 self.safety_context = closure_visitor.safety_context;
             }
+            ExprKind::Borrow { borrow_kind, arg } => {
+                let arg = match self.thir[arg].kind {
+                    ExprKind::Scope { value, .. } => value,
+                    _ => arg,
+                };
+                if let ExprKind::Field { lhs, .. } = self.thir[arg].kind {
```
Comment (LeSeulArtichaut) [inline compiler/rustc_mir_build/src/check_unsafety.rs:256]: This is incorrect and doesn't catch cases where e.g. the layout-constrained field is deeper:

```rs
#[rustc_layout_scalar_valid_range_start(1)]
struct NonZero<T>(T);

let mut foo = unsafe { NonZero((1,)) };
let a = &mut foo.0.0;
```

In fact, a place can contain all of these kinds of expressions: https://github.com/rust-lang/rust/blob/d34a3a401b4e44f289a4d5bf53da83367cbb6aa7/compiler/rustc_mir_build/src/build/expr/category.rs#L38-L44

So as long as we see these, we're still in the same place and we need to continue looking for layout-constrained fields.

File: compiler/rustc_mir_build/src/check_unsafety.rs [hunk 3]
```diff
@@ -246,11 +248,135 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
...
+                                                    this.tcx.at(pat.span),
+                                                    this.param_env,
+                                                ) =>
+                                            {
+                                                this.requires_unsafe(
+                                                    pat.span,
+                                                    BorrowOfLayoutConstrainedField,
+                                                );
+                                            }
+                                            BorrowKind::Shallow | BorrowKind::Mut { .. } => {
```
Comment (LeSeulArtichaut) [inline compiler/rustc_mir_build/src/check_unsafety.rs:336]: Also here `Shallow` is not a mutable borrow

**Review by LeSeulArtichaut — COMMENTED** [review]
Okay, sorry for the wait, here's a first review. This is certainly trickier than it looks, and I worry that we'll miss edge cases, but it's OK if we miss things in this PR because THIR unsafeck is gated under an explicitely-marked-as-unsound-do-not-enable option.

**Comment by LeSeulArtichaut** [comment]
cc @nikomatsakis you might be interested in reviewing if you have enough bandwidth

File: compiler/rustc_mir_build/src/check_unsafety.rs (hunk 2 above)
Comment (FabianWolff) [reply compiler/rustc_mir_build/src/check_unsafety.rs]: You are right; I have attempted to fix this now (your example correctly leads to an error in my implementation now). However, I think we have to stop at `Deref` expressions, which is also what the current MIR implementation seems to do, unless I am misunderstanding something:
https://github.com/rust-lang/rust/blob/308fc2322bf00b5dc12454489679de1420320f56/compiler/rustc_mir/src/transform/check_unsafety.rs#L344-L346

File: compiler/rustc_mir_build/src/check_unsafety.rs (hunk 1 above)
Comment (FabianWolff) [reply compiler/rustc_mir_build/src/check_unsafety.rs]: I have implemented your suggestion now, although with the same reservation about `Deref` as below.

File: compiler/rustc_mir_build/src/check_unsafety.rs (hunk 2 above)
Comment (LeSeulArtichaut) [reply compiler/rustc_mir_build/src/check_unsafety.rs]: Now it's too restrictive:

```rs
let foo = unsafe { NonZero((1,)) };
&mut { foo.0 }.0;
```

This compiles under MIR unsafeck because the block doesn't belong to the place and creates a new value by copying the content `(1,)`, which doesn't belong to `foo` anymore so the operation is safe. To fix this I'd expect you just need to also stop on any expression that's not in `Category::Place` (the list above).

File: compiler/rustc_mir_build/src/check_unsafety.rs [hunk 4]
```diff
@@ -246,11 +249,125 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
...
+                };
+                if let ExprKind::Field { lhs, .. } = self.thir[lhs].kind {
+                    if let ty::Adt(adt_def, _) = self.thir[lhs].ty.kind() {
+                        if (Bound::Unbounded, Bound::Unbounded)
+                            != self.tcx.layout_scalar_valid_range(adt_def.did)
+                        {
+                            self.requires_unsafe(expr.span, MutationOfLayoutConstrainedField);
+                        }
+                    }
+                }
```
Comment (LeSeulArtichaut) [inline compiler/rustc_mir_build/src/check_unsafety.rs:316]: I'd expect the same logic for assignments as for borrows, i.e. look for layout-constrained fields inside the place we assign to. Here's an example:

```rs
let mut foo = unsafe { NonZero((1,)) };
foo.0.0 = 0;
```

This is caught by the MIR unsafeck, but not by this implementation that just looks one level deep.

Comment (FabianWolff) [reply compiler/rustc_mir_build/src/check_unsafety.rs]: Fixed. I have also added test cases for all of the examples that you've mentioned. Thanks for your patience in (re-)reviewing this!

File: compiler/rustc_mir_build/src/check_unsafety.rs [hunk 5]
```diff
@@ -246,11 +292,86 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
...
+        match &*pat.kind {
+            PatKind::Variant { .. } | PatKind::Leaf { .. } => {
+                if let ty::Adt(adt_def, ..) = pat.ty.kind() {
+                    if (Bound::Unbounded, Bound::Unbounded)
+                        != self.tcx.layout_scalar_valid_range(adt_def.did)
+                    {
+                        let old_inside_adt = std::mem::replace(&mut self.inside_adt, true);
+                        visit::walk_pat(self, pat);
+                        self.inside_adt = old_inside_adt;
+                    }
```
Comment (LeSeulArtichaut) [inline compiler/rustc_mir_build/src/check_unsafety.rs:338]: So I found this case which should error but didn't:
```rs
let mut foo = unsafe { NonZero((1,)) };
match &mut foo {
    NonZero((a,)) => *a = 0,
}
```
and I couldn't understand why it didn't catch that there was a `ref mut` in there. Turns out it's just missing a `walk_pat` here and I didn't see it :sweat_smile:
```suggestion
                    } else {
                        visit::walk_pat(self, pat);
                    }
```

File: compiler/rustc_mir_build/src/check_unsafety.rs (hunk 4 above)
Comment (LeSeulArtichaut) [reply compiler/rustc_mir_build/src/check_unsafety.rs]: I'm the one who should thank you! The reason we need so many rounds of review is because I could only catch problems one by one, so thank you for bearing my inexperience with reviews :sweat_smile:

**Review by LeSeulArtichaut — COMMENTED** [review]
Okay, besides my last comment I haven't found a case where the implementation is wrong. I'll still pass this to someone with better knowledge of the language than me. In the meantime here are a few nits.

**Comment by FabianWolff** [comment]
All done, thanks again for your comments!

File: compiler/rustc_mir_build/src/check_unsafety.rs [hunk 6]
```diff
@@ -361,19 +352,140 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
...
+                }
+                _ => {}
+            },
             _ => {}
         }
         visit::walk_expr(self, expr);
     }
+
+    fn visit_pat(&mut self, pat: &Pat<'tcx>) {
```
Comment (oli-obk) [inline compiler/rustc_mir_build/src/check_unsafety.rs:400]: why did you move this function? Please either leave it where it was, or move it in a separate commit. otherwise the diff is harder to check for reviewers.

Comment (FabianWolff) [reply compiler/rustc_mir_build/src/check_unsafety.rs:400]: I didn't move it, I just resolved the merge conflicts with #85263, which happened to introduce the same function in a different place.

Comment (oli-obk) [reply compiler/rustc_mir_build/src/check_unsafety.rs]: thanks for the quick fix nonetheless

File: compiler/rustc_mir_build/src/check_unsafety.rs [hunk 7]
```diff
@@ -161,60 +209,93 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
...
+                } else {
+                    visit::walk_pat(self, pat);
+                }
+            }
+            PatKind::Binding { mode: BindingMode::ByRef(borrow_kind), ty, .. } => {
+                if self.inside_adt {
+                    if let ty::Ref(_, ty, _) = ty.kind() {
+                        match borrow_kind {
+                            BorrowKind::Shallow | BorrowKind::Shared | BorrowKind::Unique
+                                if !ty.is_freeze(self.tcx.at(pat.span), self.param_env) =>
```
Comment (oli-obk) [inline compiler/rustc_mir_build/src/check_unsafety.rs:261]: if you move the if from the guard to the arm body, you don't need the wildcard pattern and the match becomes exhaustive

File: compiler/rustc_mir_build/src/check_unsafety.rs [hunk 8]
```diff
@@ -161,60 +209,93 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
...
-                    Wild |
-                    // doesn't have any effect on semantics
-                    AscribeUserType { .. } |
-                    // creating a union literal
-                    Constant { .. } => {},
-                    Leaf { .. } | Or { .. } => {
-                        // pattern matching with a union and not doing something like v = Union { bar: 5 }
-                        self.in_union_destructure = true;
+        match &*pat.kind {
+            PatKind::Variant { .. } | PatKind::Leaf { .. } => {
```
Comment (oli-obk) [inline compiler/rustc_mir_build/src/check_unsafety.rs:236]: you can remove the variant pattern here, unions and layout constraints only work on leaf and `Or` patterns

File: compiler/rustc_mir_build/src/check_unsafety.rs [hunk 9]
```diff
@@ -161,60 +209,93 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
...
-                        unreachable!("impossible union destructuring type"),
                 }
+                visit::walk_pat(self, pat);
+            }
+            PatKind::Deref { .. } => {
+                let old_inside_adt = std::mem::replace(&mut self.inside_adt, false);
+                visit::walk_pat(self, pat);
+                self.inside_adt = old_inside_adt;
+            }
+            PatKind::Or { .. } => {
```
Comment (oli-obk) [inline compiler/rustc_mir_build/src/check_unsafety.rs:285]: why are or patterns problematic only for unions and not for layout constrained structs?

File: compiler/rustc_mir_build/src/check_unsafety.rs [hunk 10]
```diff
@@ -361,15 +442,46 @@ impl<'a, 'tcx> Visitor<'a, 'tcx> for UnsafetyVisitor<'a, 'tcx> {
...
+                    self.in_possible_lhs_union_assign = false;
+                    visit::walk_expr(self, &self.thir()[rhs]);
+                    return; // we have already visited everything by now
+                }
             }
+            ExprKind::Borrow { borrow_kind, arg } => match borrow_kind {
+                BorrowKind::Shallow | BorrowKind::Shared | BorrowKind::Unique
+                    if !self.thir[arg]
+                        .ty
+                        .is_freeze(self.tcx.at(self.thir[arg].span), self.param_env) =>
```
Comment (oli-obk) [inline compiler/rustc_mir_build/src/check_unsafety.rs:468]: also move this into the arm body

**Review by oli-obk — CHANGES_REQUESTED** [review]

File: compiler/rustc_mir_build/src/check_unsafety.rs (hunk 9 above)
Comment (FabianWolff) [reply compiler/rustc_mir_build/src/check_unsafety.rs]: Actually, I think the special case for `Or` patterns is redundant even for unions here; I have dropped it now.

I've also implemented your other suggestions; thanks for the review!
